### PR TITLE
Fidelity csv importer and common reader/investements.py updates

### PR DESCRIPTION
**Pull request by savingsandloan**
I'm still reviewing/testing a few other things in the custom fidelity_csv importer before pushing it, but here's the common-code changes so far.

### CODE REVIEW

**Comment by savingsandloan** [comment]
Still fixing some things. Is there a certain linter setup (e.g. black configuration) used?

**Comment by redstreet** [comment]
@savingsandloan, apologies I hadn't responded yet. Thanks for the ping on the group. Could you please let me know when you've had a chance to resolve the conflicts, and address the comments above, and I'll be happy to review?

**Comment by savingsandloan** [comment]
@redstreet no problem, started merging, doing a bit more testing and saw some differences in how we were approaching generate_transfer_entry - I'm currently [erring toward](https://github.com/redstreet/beancount_reds_importers/pull/17/commits/043a3cfb5bffab331fe99c046cc50da63f311169#diff-975db43710a6407edfe8411adae4995ba72a7bcf21b1514e01d5638b5335b4b9R247-R277) what I had which was closer to your original code. 

The other thing I'm realizing is that with a configurable `cash_account`, how I'm approaching configurable `transfer_accounts` also needs to be aware of both `cash_account` and `main_account` to handle cash-cash transfers versus commodity transfers/purchases.

**Comment by redstreet** [comment]
Left a few comments in code. Here are some overall comments:

### Code location
- bond related code: would you mind very briefly describing what fundamental problem is being solved? Is it just the four metadata items you show in #12? I imagine most bond code should be in a separate bond library, usable by all importers

- transfer related code: you should be able to accomplish all you need in a custom importer, rather than putting it in `investments.py`. beancount_reds_importers's philosophy is to keep the core simple enough that one can understand and mostly verify it by mere inspection

### Code philosophy
- how can the philosophy of the code in the PR match the core premise of beancount_reds_importers, which is simplicity?

### Code comments
my two cents: IMHO, the parsing code is brittle, easily prone to breakage when fidelity makes small changes (which in my experience is generally common). I'd recommend looking for ways to generalize and simplify all the cases into just a handful

**Comment by savingsandloan** [comment]
> Left a few comments in code. Here are some overall comments:
> ### Code location
> 
>     * bond related code: would you mind very briefly describing what fundamental problem is being solved? Is it just the four metadata items you show in [Fix handling for readers when available_cash is not present. #12](https://github.com/redstreet/beancount_reds_importers/pull/12)? I imagine most bond code should be in a separate bond library, usable by all importers
> 
The custom fidelity csv importer currently does:
- Correcting from par value to actual purchase price, which is likely specific to fidelity
- Detection of different types of bond purchases based on description, again, likely specific to fidelity
- Generation of metadata specific to bonds (term length, bond type, coupon rate, expiration date, ...)

For the last point I agree there should be a common bond library to facilitate generating appropriate metadata. I can open an issue for that to be done later.

>     * transfer related code: you should be able to accomplish all you need in a custom importer, rather than putting it in `investments.py`. beancount_reds_importers's philosophy is to keep the core simple enough that one can understand and mostly verify it by mere inspection

You are correct. The changes I was mentioning were more of a self-note and only impact the fidelty csv importer, not investments.py.

> 
> 
> ### Code philosophy
> 
>     * how can the philosophy of the code in the PR match the core premise of beancount_reds_importers, which is simplicity?
> 

I think for most of the current changes to the non-custom-importer sections, I already tried to largely just extend the structures and fix a few areas. This said, there are things that aren't as simple as expected in the current structure of `investments.py`. Some examples are:
- In `investments.py` there's currently `target_account_map` which isn't a guaranteed map as there's different types of transfer scenarios (_cash to shares_, _shares to shares_, _cash to cash_) as coded in `generate_transfer_entry` which'll determine the sourcing and targeting accounts. 
- `investments.py` in it's current form is crafted around the OFX format. This is generally fine, but leads to questions around its conventions, such as when should one use `unit`, `total`, or `amount`.
   - Transfer_types: I held off on it for the moment, but there could be improvement in the naming of and use of where I had the following:
      ```
        self.transfer_unit_types = ['transfer']
        self.transfer_amount_types = [
            'other', 'credit', 'debit', 'dep', 'cash']
        self.transfer_other_types = [
            'income', 'dividends', 'capgains_st', 'capgains_lt']
      ```
      where `transfer_other_types` could be renamed to reflect the cash (ofx `total`?) intention of those, and used where string lists are currently used (e.g. `generate_transfer_entry`)

As for the custom importer, there's the following things that could be migrated into something more common:
   - Common library for bond-related actions - can file an issue to do this later.
   - A common way of validating+handling CUSIPs when there isn't a ticker symbol (happens with bonds and likely other instruments) - can file an issue to do this later. 
   - Some commenting of why certain things are done to help future maintainers - for example why `__associate_transaction_values` is currently needed to migrate say, certain values in a csv file's `total` field to be in the `amount` field to follow the ofx-oriented convention of `investments.py`. This is related to the point above about _Transfer_types_

> 
> ### Code comments
> 
>     * my two cents: IMHO, the parsing code is brittle, easily prone to breakage when fidelity makes small changes (which in my experience is generally common). I'd recommend looking for ways to generalize and simplify all the cases into just a handful

Fair point, though will note the importer is based on what I've been using for a couple years to cover various cases and am still working to find a tradeoff between simplicity (where a user may have to correct things later) and being relatively comprehensive. Again, as mentioned while there's some brittleness in the current code I can file issues to break some aspects of it into more common libraries later.

To summarize, my proposal is to:

 - I'll add couple more commits to this PR to DRY/fix some namings and finish custom-importer aspects around transfers.
 - File issue as a reminder to make a bond-specific library in the future.
 - File issue as reminder to handle CUSIPs when there isn't a ticker symbol.

**Comment by redstreet** [comment]
Thanks for the helpful responses! I have a better sense of of the issues now.

My concern with this PR is its scope. Here are the components it has:
1. basic fidelity csv importer
1. advanced fidelity csv parsing to support components below
1. cusip handling
1. metadata additions: account open/close, commodity
1. bond metadata
1. transfers and deduping

Each brings up interesting questions around code philosophy and code architecture. Trying to address and handle all of them at once is unwieldy and virtually impossible. It would be much easier if we could handle each of these separately, and merge them in if we see fit, one at a time, before moving on to the next one. Do let me know if you have concerns. There might be just a little bit of work involved in splitting these, but it will let us focus on the issues much better, and keep the code architecture consistent, and keep the quality of `master` high.

If that sounds good to you, I'd suggest starting with (1) above, which would be a basic fidelity csv importer, which looks more or less very close to the [schwab_csv](https://github.com/redstreet/beancount_reds_importers/tree/master/beancount_reds_importers/schwab_csv) importer.

**Comment by savingsandloan** [comment]
Okay, I'll think about it, may take me some time (busy between updating finances, taxes, finding work) to re-approach this. I understand the concern, just trying to think about the structured approach alongside current usage/testing for personal needs.

**Comment by redstreet** [comment]
Of course, that makes sense. Happy to help in any way I can, just reach out.

**Comment by redstreet** [comment]
- generate_transfer_entry bug: yes, please to file a bug/PR if you can confirm it.
- metadata additions: I'm assuming this is just a line or two in `investments.py`

I'm happy to help you get started on the fidelity csv importer by doing 1a above (a brokerage importer which looks more or less very close to the [schwab_csv](https://github.com/redstreet/beancount_reds_importers/tree/master/beancount_reds_importers/schwab_csv) importer). That might help you complete 1b/1c, and then move on to the remaining.

**Comment by redstreet** [comment]
I've written a basic fidelity brokerage importer. See 848613e. Try running the test file in it. I'm hoping this is helpful in showing how the column parsing and transaction type determining code can be built.

A separate (?) HSA importer should be trivial to build based on this. A separate, third banking importer should be even easier to build. I'd keep these separate since they involve different types of transactions.

Once you have these built, you can make calls to other common components (cusip, bond, dedupe etc.) with just a couple lines of code to the importer. If importers get any bigger than say, a 100 lines, that's a good indication that we need to put more thought into the code architecture to keep things simple and maintainable.

**Comment by redstreet** [comment]
Renamed master to main, which caused this to get closed. We can continue the discussion here, and you're welcome to submit separate PRs for the components above as we are discussing. Thanks!

**Comment by savingsandloan** [comment]
Note: I've moved my initial 'prototype' branch to (fidelity_csv_old)[https://github.com/savingsandloan/beancount_reds_importers/commits/fidelity_csv_old]

**Comment by tbm** [comment]
@savingsandloan are you planning on creating new PRs?  Would be great to see this in.

**Comment by savingsandloan** [comment]
@tbm yes, been juggling a few things but slowly coming back to this.

@redstreet I know we've been talking about some finer points on #26 , but have you had any thoughts on #28 ? The structure of that PR actually influences another branch I haven't made a PR for yet around creating accounts/commodity postings.

**Comment by redstreet** [comment]
@savingsandloan, yikes, I left feedback a long time ago but forgot to hit submit. Apologies. Done.
